File: packages/util/bridge-protocol/src/framing.ts
/**
 * LSP-compatible Content-Length framing for product-bridge JSON messages.
 * @module @deepseek-ai/dsh-bridge-protocol
 */

import { parseBridgeMessage } from './message.ts'
import type { BridgeMessage, FrameErrorCode } from './types.ts'
import { DEFAULT_MAX_FRAME_SIZE } from './types.ts'

const SEPARATOR = Buffer.from('\r\n\r\n')

/** Framing failure with a stable code. */
export class FrameError extends Error {
  override readonly name = 'FrameError'

  /**
   * @param code - stable framing failure class.
   * @param message - diagnostic retained as the Error message.
   */
  constructor(readonly code: FrameErrorCode, message: string) {
    super(message)
  }
}

/**
 * Encode one message using the LSP-compatible Content-Length carrier.
 * @param message - logical bridge message.
 * @returns the complete frame bytes.
 */
export function encodeFrame(message: BridgeMessage): Buffer {
  const body = Buffer.from(JSON.stringify(message), 'utf8')
  const header = Buffer.from(`Content-Length: ${body.length}\r\n\r\n`, 'utf8')
  return Buffer.concat([header, body])
}

/**
 * Incremental decoder that accepts arbitrary transport chunk boundaries.
 */
export class FrameDecoder {
  private buffer = Buffer.alloc(0)

  /**
   * @param maxFrameSize - maximum admitted body size in bytes.
   */
  constructor(private readonly maxFrameSize: number = DEFAULT_MAX_FRAME_SIZE) {}

  /**
   * Add bytes and return every complete message currently available.
   * @param bytes - next transport chunk.
   * @returns complete messages decoded from the accumulated buffer.
   */
  push(bytes: Uint8Array): BridgeMessage[] {
    this.buffer = this.buffer.length === 0 ? Buffer.from(bytes) : Buffer.concat([this.buffer, bytes])
    const messages: BridgeMessage[] = []
    for (;;) {
      const separator = this.buffer.indexOf(SEPARATOR)
      if (separator < 0) {
        if (this.buffer.length > 4096) throw new FrameError('header-too-large', 'frame header exceeds the 4096 byte limit')
        break
      }
      if (separator > 4096) throw new FrameError('header-too-large', 'frame header exceeds the 4096 byte limit')
      let header: string
      try {
        header = this.buffer.subarray(0, separator).toString('utf8')
      } catch {
        throw new FrameError('header-encoding', 'frame header is not valid UTF-8')
      }
      const contentLength = parseContentLength(header)
      if (contentLength > this.maxFrameSize) {
        throw new FrameError('too-large', `frame exceeds the ${this.maxFrameSize} byte limit`)
      }
      const bodyStart = separator + 4
      const frameEnd = bodyStart + contentLength
      if (this.buffer.length < frameEnd) break
      const body = this.buffer.subarray(bodyStart, frameEnd).toString('utf8')
      this.buffer = this.buffer.subarray(frameEnd)
      let parsed: unknown
      try {
        parsed = JSON.parse(body)
      } catch (error) {
        throw new FrameError('json', `frame JSON is invalid: ${error instanceof Error ? error.message : String(error)}`)
      }
      messages.push(parseBridgeMessage(parsed))
    }
    return messages
  }

  /**
   * Number of bytes waiting for a complete frame.
   * @returns the unread buffer length.
   */
  get bufferedLen(): number {
    return this.buffer.length
  }
}

function parseContentLength(header: string): number {
  let parsed: number | undefined
  for (const line of header.split('\r\n')) {
    const split = line.indexOf(':')
    if (split < 0) continue
    if (line.slice(0, split).toLowerCase() !== 'content-length') continue
    if (parsed !== undefined) throw new FrameError('duplicate-content-length', 'frame header contains more than one Content-Length')
    const value = Number.parseInt(line.slice(split + 1).trim(), 10)
    if (!Number.isInteger(value) || value < 0) {
      throw new FrameError('invalid-content-length', 'frame Content-Length is invalid')
    }
    parsed = value
  }
  if (parsed === undefined) throw new FrameError('missing-content-length', 'frame header is missing Content-Length')
  return parsed
}

File: native/dsh/crates/dsh-bridge-protocol/src/framing.rs
use crate::BridgeMessage;
use serde_json::Error as JsonError;
use std::io::Write;
use thiserror::Error;

/// Errors raised while decoding or writing Content-Length frames.
#[derive(Debug, Error)]
pub enum FrameError {
    #[error("frame header is not valid UTF-8")]
    HeaderEncoding,
    #[error("frame header is missing Content-Length")]
    MissingContentLength,
    #[error("frame header contains more than one Content-Length")]
    DuplicateContentLength,
    #[error("frame header exceeds the 4096 byte limit")]
    HeaderTooLarge,
    #[error("frame Content-Length is invalid")]
    InvalidContentLength,
    #[error("frame exceeds the {max} byte limit")]
    TooLarge { max: usize },
    #[error("frame JSON is invalid: {0}")]
    Json(#[from] JsonError),
    #[error("frame write failed: {0}")]
    Io(#[from] std::io::Error),
}

/// Default maximum logical frame size for the temporary JSON carrier.
pub const DEFAULT_MAX_FRAME_SIZE: usize = 16 * 1024 * 1024;

/// Encodes one message using the LSP-compatible Content-Length carrier.
pub fn encode_frame(message: &BridgeMessage) -> Result<Vec<u8>, FrameError> {
    let body = serde_json::to_vec(message).map_err(FrameError::Json)?;
    let header = format!("Content-Length: {}\r\n\r\n", body.len());
    let mut frame = Vec::with_capacity(header.len() + body.len());
    frame.extend_from_slice(header.as_bytes());
    frame.extend_from_slice(&body);
    Ok(frame)
}

/// Writes one encoded frame without mixing diagnostics into stdout.
pub fn write_frame<W: Write>(writer: &mut W, message: &BridgeMessage) -> Result<(), FrameError> {
    writer.write_all(&encode_frame(message)?)?;
    writer.flush()?;
    Ok(())
}

/// Incremental decoder that accepts arbitrary transport chunk boundaries.
pub struct FrameDecoder {
    buffer: Vec<u8>,
    max_frame_size: usize,
}

impl Default for FrameDecoder {
    fn default() -> Self {
        Self::new(DEFAULT_MAX_FRAME_SIZE)
    }
}

impl FrameDecoder {
    /// Creates a decoder with an explicit body-size limit.
    pub fn new(max_frame_size: usize) -> Self {
        Self {
            buffer: Vec::new(),
            max_frame_size,
        }
    }

    /// Adds bytes and returns every complete message currently available.
    pub fn push(&mut self, bytes: &[u8]) -> Result<Vec<BridgeMessage>, FrameError> {
        self.buffer.extend_from_slice(bytes);
        let mut messages = Vec::new();
        loop {
            let Some(separator) = self
                .buffer
                .windows(4)
                .position(|window| window == b"\r\n\r\n")
            else {
                if self.buffer.len() > 4096 {
                    return Err(FrameError::HeaderTooLarge);
                }
                break;
            };
            if separator > 4096 {
                return Err(FrameError::HeaderTooLarge);
            }
            let header = std::str::from_utf8(&self.buffer[..separator])
                .map_err(|_| FrameError::HeaderEncoding)?;
            let content_length = parse_content_length(header)?;
            if content_length > self.max_frame_size {
                return Err(FrameError::TooLarge {
                    max: self.max_frame_size,
                });
            }
            let body_start = separator + 4;
            let frame_end = body_start + content_length;
            if self.buffer.len() < frame_end {
                break;
            }
            let body: Vec<u8> = self.buffer[body_start..frame_end].to_vec();
            self.buffer.drain(..frame_end);
            messages.push(serde_json::from_slice(&body)?);
        }
        Ok(messages)
    }

    /// Returns the number of bytes waiting for a complete frame.
    pub fn buffered_len(&self) -> usize {
        self.buffer.len()
    }
}

fn parse_content_length(header: &str) -> Result<usize, FrameError> {
    let mut parsed = None;
    for line in header.split("\r\n") {
        let Some((name, value)) = line.split_once(':') else {
            continue;
        };
        if name.eq_ignore_ascii_case("content-length") {
            if parsed.is_some() {
                return Err(FrameError::DuplicateContentLength);
            }
            parsed = Some(
                value
                    .trim()
                    .parse::<usize>()
                    .map_err(|_| FrameError::InvalidContentLength)?,
            );
        }
    }
    parsed.ok_or(FrameError::MissingContentLength)
}
